rust-runtime/aws-smithy-runtime-api/src/client/interceptors/context/wrappers.rs

```diff
@@ -0,0 +1,207 @@
+pub struct BeforeDeserializationInterceptorContextRef<'a, I, O, E> {
+    inner: &'a InterceptorContext<I, O, E>,
+}
+
+impl<'a, I, O, E> From<&'a InterceptorContext<I, O, E>>
+    for BeforeDeserializationInterceptorContextRef<'a, I, O, E>
+{
+    fn from(inner: &'a InterceptorContext<I, O, E>) -> Self {
+        Self { inner }
+    }
+}
+
+impl<'a, I, O, E> BeforeDeserializationInterceptorContextRef<'a, I, O, E> {
+    pub fn input(&self) -> &I {
+        self.inner.input()
+    }
+
+    /// Retrieve the transmittable request for the operation being invoked.
+    pub fn request(&self) -> &Request {
+        self.inner.request()
+    }
+
+    /// Returns the response.
+    pub fn response(&self) -> &Response {
+        self.inner.response()
+    }
+}
+
+pub struct BeforeDeserializationInterceptorContextMut<'a, I, O, E> {
+    inner: &'a mut InterceptorContext<I, O, E>,
+}
+
+impl<'a, I, O, E> From<&'a mut InterceptorContext<I, O, E>>
+    for BeforeDeserializationInterceptorContextMut<'a, I, O, E>
+{
+    fn from(inner: &'a mut InterceptorContext<I, O, E>) -> Self {
+        Self { inner }
+    }
+}
+
+impl<'a, I, O, E> BeforeDeserializationInterceptorContextMut<'a, I, O, E> {
+    pub fn input_mut(&mut self) -> &mut I {
+        self.inner.input_mut()
+    }
+    /// Retrieve the transmittable request for the operation being invoked.
+    pub fn request_mut(&mut self) -> &mut Request {
+        self.inner.request_mut()
+    }
+
+    /// Returns a mutable reference to the response.
+    pub fn response_mut(&mut self) -> &mut Response {
+        self.inner.response_mut()
+    }
+}
+
+pub struct AfterDeserializationInterceptorContextRef<'a, I, O, E> {
+    inner: &'a InterceptorContext<I, O, E>,
+}
+
+impl<'a, I, O, E> From<&'a InterceptorContext<I, O, E>>
+    for AfterDeserializationInterceptorContextRef<'a, I, O, E>
+{
+    fn from(inner: &'a InterceptorContext<I, O, E>) -> Self {
+        Self { inner }
+    }
+}
+
+impl<'a, I, O, E> AfterDeserializationInterceptorContextRef<'a, I, O, E> {
+    pub fn input(&self) -> &I {
+        self.inner.input()
+    }
+
+    /// Retrieve the transmittable request for the operation being invoked.
+    pub fn request(&self) -> &Request {
+        self.inner.request()
+    }
+
+    /// Returns the response.
+    pub fn response(&self) -> &Response {
+        self.inner.response()
+    }
+
+    /// Returns the deserialized output or error.
+    pub fn output_or_error(&self) -> Result<&O, &E> {
+        self.inner.output_or_error()
+    }
+}
+
+pub struct AfterDeserializationInterceptorContextMut<'a, I, O, E> {
+    inner: &'a mut InterceptorContext<I, O, E>,
+}
+
+impl<'a, I, O, E> From<&'a mut InterceptorContext<I, O, E>>
+    for AfterDeserializationInterceptorContextMut<'a, I, O, E>
+{
+    fn from(inner: &'a mut InterceptorContext<I, O, E>) -> Self {
+        Self { inner }
+    }
+}
+
+impl<'a, I, O, E> AfterDeserializationInterceptorContextMut<'a, I, O, E> {
+    pub fn input_mut(&mut self) -> &mut I {
```

Comment:
We should make a macro for each accessor method so that we don't have to repeat the documentation everywhere and worry about consistency. The `-Mut` impls should also have the non-mut functions, and a macro will probably make that easier.